presto-accumulo/src/main/java/com/facebook/presto/accumulo/AccumuloMetadata.java

```diff
@@ -0,0 +1,427 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+
+import javax.inject.Inject;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.accumulo.AccumuloErrorCode.ACCUMULO_TABLE_EXISTS;
+import static com.facebook.presto.accumulo.AccumuloErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.accumulo.Types.checkType;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Presto metadata provider for Accumulo. Responsible for creating/dropping/listing tables, schemas,
+ * columns, all sorts of goodness. Heavily leverages {@link AccumuloClient}
+ */
+public class AccumuloMetadata
+        implements ConnectorMetadata
+{
+    private final String connectorId;
+    private final AccumuloClient client;
+
+    private final AtomicReference<Runnable> rollbackAction = new AtomicReference<>();
+
+    @Inject
+    public AccumuloMetadata(AccumuloConnectorId connectorId, AccumuloClient client)
+    {
+        this.connectorId = requireNonNull(connectorId, "connectorId is null").toString();
+        this.client = requireNonNull(client, "client is null");
+    }
+
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session,
+            ConnectorTableMetadata tableMetadata, Optional<ConnectorNewTableLayout> layout)
+    {
+        checkNoRollback();
+
+        SchemaTableName stName = tableMetadata.getTable();
+        AccumuloTable table = client.createTable(tableMetadata);
+
+        AccumuloTableHandle handle = new AccumuloTableHandle(connectorId, stName.getSchemaName(), stName.getTableName(),
+                table.getRowId(), table.isExternal(), table.getSerializerClassName(),
+                table.getScanAuthorizations());
+
+        setRollback(() -> rollbackCreateTable(table));
+
+        return handle;
+    }
+
+    @Override
+    public void finishCreateTable(ConnectorSession session,
+            ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments)
+    {
+        checkType(tableHandle, AccumuloTableHandle.class, "tableHandle");
+        clearRollback();
+    }
+
+    private void rollbackCreateTable(AccumuloTable table)
+    {
+        client.dropTable(table);
+    }
+
+    @Override
+    public void createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
+    {
+        client.createTable(tableMetadata);
+    }
+
+    @Override
+    public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        AccumuloTableHandle th = checkType(tableHandle, AccumuloTableHandle.class, "table");
+        client.dropTable(client.getTable(th.toSchemaTableName()));
```

Comment:
`client.getTable` may return `null`. You're not checking for nullability of `AccumuloTable` in `AccumuloClient.dropTable`.